### DISCUSSION TITLE
Is it possible to create an common psalm config that can be extended on?

### DISCUSSION THREAD

**Discussion by Ilyes512** [vimeo/psalm#6276] [General]
So I have multiple packages that I maintain and I wan't to create a common psalm config. In the docs I found a way of splitting up configs and including them, but I did not find a way to extend whatever is included.

So to use the same example from https://psalm.dev/docs/running_psalm/configuration I want to be able to do something like:

**common-files.xml**
```xml
<?xml version="1.0" encoding="UTF-8"?>
<projectFiles xmlns="https://getpsalm.org/schema/config">
    <file name="Bar.php" />
    <file name="Bat.php" />
</projectFiles>
```

**psalm.xml**
```xml
<?xml version="1.0"?>
<psalm
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xmlns="https://getpsalm.org/schema/config"
    xsi:schemaLocation="https://getpsalm.org/schema/config vendor/vimeo/psalm/config.xsd"
    xmlns:xi="http://www.w3.org/2001/XInclude"
>
    <xi:include href="common-files.xml"/>
    <!-- Add an extra file: -->
    <projectFiles>
        <file name="anotherFile.php" />
    </projectFiles>
</psalm>
```

Maybe I am on the wrong path? Maybe there is a way to point to 2 configs (or even more) at the same time and have them (recursively) "merge"?

**Comment by weirdan** [comment]
> Maybe there is a way to point to 2 configs (or even more) at the same time and have them (recursively) "merge"?

No, Psalm supports a single config only. But you can include multiple elements with XInclude, by using the `xpointer` attribute:

```xml
<?xml version="1.0"?>
<psalm
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xmlns="https://getpsalm.org/schema/config"
    xsi:schemaLocation="https://getpsalm.org/schema/config vendor/vimeo/psalm/config.xsd"
    xmlns:xi="http://www.w3.org/2001/XInclude"
>
    <projectFiles>
        <xi:include 
            href="common-files.xml" 
            xpointer="xmlns(ps=https://getpsalm.org/schema/config)xpointer(/ps:projectFiles/*)"
        />
        <file name="anotherFile.php" />
    </projectFiles>
</psalm>
```

**Reply by gmessier** [reply]
You may also like those PRs that I did to improve capabilities to have a configuration built from many sources.
My team also want to have a base configuration that will be applied to all our packages. 

#6516
#6519
